File: 0077 Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res, combo = [], []
        if n <= 0 or k <= 0: return res            
        self.findCombos(res, n, k, 1, combo)
        return res
        
            
    def findCombos(self, res: List[List[int]], n: int, k: int, i: int, combo: List[int]) -> None:
        if k == 0:
            res.append(combo.copy())
        elif i + k - 1 <= n:
            combo.append(i)
            self.findCombos(res, n, k - 1, i + 1, combo)
            combo.pop()
            self.findCombos(res, n, k, i + 1, combo)

''' Don't use it in the interview!
from itertools import combinations
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        return combinations(range(1, n+1), k) 
'''

File: 0077 Combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        ArrayList<Integer> combo = new ArrayList<Integer>();
        if (n <= 0 || k <= 0) return res;
        findCombos(res, n, k, 1, combo);
        return res;
    }
    
    public void findCombos(List<List<Integer>> res, int n, int k, int i, ArrayList<Integer> combo) {
        if (k == 0) {
            res.add(new ArrayList<>(combo));
        } else if (i + k - 1 <= n) {
            combo.add(i);
            findCombos(res, n, k - 1, i + 1, combo);
            combo.remove(combo.size() - 1);
            findCombos(res, n, k, i + 1, combo);
        }
    }
}